library/Garden/Events/ResourceEvent.php

```diff
@@ -0,0 +1,86 @@
+<?php
+/**
+ * @copyright 2009-2020 Vanilla Forums Inc.
+ * @license GPL-2.0-only
+ */
+
+namespace Garden\Events;
+
+/**
+ * An event affecting a specific resource.
+ */
+abstract class ResourceEvent {
+
+    /** A resource has been removed. */
+    const ACTION_DELETE = "delete";
+
+    /** A resource has been created. */
+    const ACTION_INSERT = "insert";
+
+    /** An existing resource has been updated. */
+    const ACTION_UPDATE = "update";
+
+    /** @var string */
+    protected $action;
+
+    /** @var array */
+    protected $payload;
+
+    /** @var string */
+    protected $type;
+
+    /**
+     * Create the event.
+     *
+     * @param string $action
+     * @param array $payload
+     */
+    public function __construct(string $action, array $payload) {
+        $this->action = $action;
+        $this->payload = $payload;
+        $this->type = $this->typeFromClass();
+    }
+
+    /**
+     * Get the event action.
+     *
+     * @return string|null
+     */
+    public function getAction(): ?string {
+        return $this->action;
+    }
+
+    /**
+     * Get the event payload.
+     *
+     * @return array|null
+     */
+    public function getPayload(): ?array {
+        return $this->payload;
+    }
+
+    /**
+     * Get the event type.
+     *
+     * @return string|null
+     */
+    public function getType(): ?string {
```

Comment:
I think this should just be `string` to avoid edge cases. Also, the constructor makes it a string.